api/git/azure.go

```diff
@@ -294,3 +396,189 @@ func getVersionType(name string) string {
+		} else if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusNonAuthoritativeInfo {
+			return nil, ErrAuthenticationFailure
+		}
+		return nil, fmt.Errorf("failed to list refs url with a status \"%v\"", resp.Status)
+	}
+
+	var refs struct {
+		Value []azureRef
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&refs); err != nil {
+		return nil, errors.Wrap(err, "could not parse Azure refs response")
+	}
+
+	var ret []string
+	for _, value := range refs.Value {
+		if value.Name == "HEAD" {
+			continue
+		}
+		ret = append(ret, value.Name)
+	}
+
+	if a.cacheEnabled {
+		a.repoRefCache.Add(opt.repositoryUrl, ret)
+	}
+	return ret, nil
+}
+
+func (a *azureClient) listFiles(ctx context.Context, opt option) ([]string, error) {
+	var (
+		allPaths    []string
+		filteredRet []string
+		refs        []string
+		err         error
+		success     bool
+		ok          bool
+		cache       interface{}
+	)
+
+	repoKey := generateCacheKey(opt.repositoryUrl, opt.referenceName)
+	if a.repoFileCache != nil {
+		cache, ok = a.repoFileCache.Get(repoKey)
+		if ok {
+			treeCache, success := cache.([]string)
+			if success {
+				for _, path := range treeCache {
+					if matchExtensions(path, opt.extensions) {
+						filteredRet = append(filteredRet, path)
+					}
+				}
+				return filteredRet, nil
+			}
+		}
+	}
+
+	// Check if the reference exists
+	if a.repoRefCache != nil {
+		cache, ok = a.repoRefCache.Get(opt.repositoryUrl)
+		if ok {
+			refs, success = cache.([]string)
+		}
+	}
+	if !ok || !success {
+		opt := option{
+			repositoryUrl: opt.repositoryUrl,
+			username:      opt.username,
+			password:      opt.password,
+		}
+		refs, err = a.listRefs(ctx, opt)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	matchedRef := false
+	for _, ref := range refs {
+		if ref == opt.referenceName {
+			matchedRef = true
+			break
+		}
+	}
+	if !matchedRef {
+		return nil, ErrRefNotFound
+	}
+
+	//
+	rootItem, err := a.getRootItem(ctx, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := parseUrl(opt.repositoryUrl)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to parse url")
+	}
+
+	listTreeUrl, err := a.buildTreeUrl(config, rootItem.ObjectID)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to build list tree url")
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "GET", listTreeUrl, nil)
+	if opt.username != "" || opt.password != "" {
+		req.SetBasicAuth(opt.username, opt.password)
+	} else if config.username != "" || config.password != "" {
+		req.SetBasicAuth(config.username, config.password)
+	}
+
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to create a new HTTP request")
+	}
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to make an HTTP request")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to list tree url with a status \"%v\"", resp.Status)
+	}
```

Comment:
What's the benefit of first doing a request to get a sha1 of the top commit and then fetching all files for a given commit instead of downloading the repo? Aren't we doing the exact same thing, just with an extra http request and a whole bunch of new code?
Similar is true for the git implementation.

Reply:
As mentioned above, azure listTree API requires to specify the `sha1` value. And also listTree API will only return the filenames instead of fetching all files.

Reply:
Right, I've missed the fact that `listTree` call doesn't fetch the content, now it makes more sense to use it 👍